File: android/app/src/main/java/com/voicerewriter/textproc/Capitalizer.kt
package com.voicerewriter.textproc

/**
 * Deterministic sentence capitalization: capitalize the first letter of the text,
 * the first letter after a sentence terminator (`.` `!` `?` followed by whitespace),
 * and the first letter after a newline. Never lowercases anything, so existing
 * casing ("iPhone", proper nouns) is preserved. Mid-token dots ("fast.ai", "3.14")
 * are not treated as terminators because they aren't followed by whitespace.
 *
 * Inferred *commas* / sentence boundaries are out of scope here (that's the
 * fine-tuned LLM's job); this only fixes capitalization at obvious boundaries.
 */
object Capitalizer {

    // Leading characters to skip over when looking for the first letter of a sentence.
    private val leadingMarks = setOf('"', '“', '\'', '‘', '(', '[')

    fun capitalizeSentences(text: String): String {
        if (text.isEmpty()) return text
        val sb = StringBuilder(text)
        var capNext = true
        for (i in sb.indices) {
            val c = sb[i]
            when {
                c == '\n' -> capNext = true
                c.isWhitespace() || c in leadingMarks -> { /* keep waiting for the first letter */ }
                capNext && c.isLetter() -> { sb[i] = c.uppercaseChar(); capNext = false }
                else -> capNext = false // sentence content has begun
            }
            // A terminator starts a new sentence only when followed by whitespace/end.
            if (c == '.' || c == '!' || c == '?') {
                val next = if (i + 1 < sb.length) sb[i + 1] else ' '
                if (next.isWhitespace()) capNext = true
            }
        }
        return sb.toString()
    }
}

File: macos/OpenWisprCore/Sources/OpenWisprCore/Capitalizer.swift
import Foundation

/// Deterministic sentence capitalization. Port of the Kotlin `Capitalizer`.
enum Capitalizer {

    // Leading characters to skip over when looking for the first letter of a sentence.
    private static let leadingMarks: Set<Character> = ["\"", "\u{201C}", "'", "\u{2018}", "(", "["]

    static func capitalizeSentences(_ text: String) -> String {
        if text.isEmpty { return text }
        var sb = Array(text)
        var capNext = true
        for i in sb.indices {
            let c = sb[i]
            if c == "\n" {
                capNext = true
            } else if c.isWhitespace || leadingMarks.contains(c) {
                // keep waiting for the first letter
            } else if capNext && c.isLetter {
                sb[i] = uppercaseChar(c); capNext = false
            } else {
                capNext = false // sentence content has begun
            }
            // A terminator starts a new sentence only when followed by whitespace/end.
            if c == "." || c == "!" || c == "?" {
                let next: Character = (i + 1 < sb.count) ? sb[i + 1] : " "
                if next.isWhitespace { capNext = true }
            }
        }
        return String(sb)
    }

    /// Mirrors Kotlin `Char.uppercaseChar()` — single-character uppercase.
    private static func uppercaseChar(_ c: Character) -> Character {
        let up = String(c).uppercased()
        return up.first ?? c
    }
}
